presto-main/src/main/java/com/facebook/presto/sql/planner/LambdaCaptureDesugaringRewriter.java

```diff
@@ -0,0 +1,139 @@
+import com.facebook.presto.sql.tree.BindExpression;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.ExpressionRewriter;
+import com.facebook.presto.sql.tree.ExpressionTreeRewriter;
+import com.facebook.presto.sql.tree.LambdaArgumentDeclaration;
+import com.facebook.presto.sql.tree.LambdaExpression;
+import com.facebook.presto.sql.tree.SymbolReference;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public class LambdaCaptureDesugaringRewriter
+{
+    private final Map<Symbol, Type> symbolTypes;
+    private final SymbolAllocator symbolAllocator;
+
+    public LambdaCaptureDesugaringRewriter(Map<Symbol, Type> symbolTypes, SymbolAllocator symbolAllocator)
+    {
+        this.symbolTypes = requireNonNull(symbolTypes, "symbolTypes is null");
+        this.symbolAllocator = requireNonNull(symbolAllocator, "symbolAllocator is null");
+    }
+
+    public Expression rewrite(Expression expression)
+    {
+        return ExpressionTreeRewriter.rewriteWith(new Visitor(), expression, new Context());
+    }
+
+    private static Expression replaceSymbols(Expression expression, ImmutableMap<Symbol, Symbol> symbolMapping)
+    {
+        return ExpressionTreeRewriter.rewriteWith(
+                new ExpressionRewriter<Void>() {
+                    @Override
+                    public Expression rewriteSymbolReference(SymbolReference node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+                    {
+                        Symbol mapTo = symbolMapping.get(new Symbol(node.getName()));
+                        if (mapTo == null) {
+                            return node;
+                        }
+                        return mapTo.toSymbolReference();
+                    }
+                },
+                expression);
+    }
+
+    public class Visitor
+            extends ExpressionRewriter<Context>
+    {
+        @Override
+        public Expression rewriteLambdaExpression(LambdaExpression node, Context context, ExpressionTreeRewriter<Context> treeRewriter)
+        {
+            HashSet<Symbol> referencedSymbols = new HashSet<>();
+            Expression rewrittenBody = treeRewriter.rewrite(node.getBody(), context.withReferencedSymbols(referencedSymbols));
+
+            ImmutableList<Symbol> lambdaArguments = node.getArguments().stream()
+                    .map(LambdaArgumentDeclaration::getName)
+                    .map(Symbol::new)
+                    .collect(toImmutableList());
+            referencedSymbols.removeAll(lambdaArguments);
+
+            // x -> f(x, captureSymbol)    will be rewritten into
+            // "$internal$bind"(captureSymbol, (extraSymbol, x) -> f(x, extraSymbol))
```

Comment:
Is `"$internal$bind"(captureSymbol, (extraSymbol, x) -> body)` a function:
`(x) -> body[s/extraSymbol/captureSymbol/]`?
Does it have signature:
```
$internal$bind(X, (X, Y) -> Z): Y->Z
```

If that is the case then what is the advantage of having intermittent `BindExpression` instead of directly generating a bytecode for original lambda: `x -> f(x, captureSymbol)`? Could you explain the reasoning? Please add the description as part of `BindExpression` Javadoc